File: JavaScript/src/add_binary.js
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
function addBinary(a, b) {
  const len_a = a.length;
  const len_b = b.length;

  let i = len_a - 1;
  let j = len_b - 1;
  let carry = 0;

  let result = "";
  while (j >= 0 || j >= 0 || carry > 0) {
    let bit1 = i >= 0 ? Number(a[i--]) : 0;
    let bit2 = j >= 0 ? Number(b[j--]) : 0;
    let sum = bit1 + bit2 + carry;
    result += (sum % 2).toString();
    carry = Math.floor(sum / 2);
  }

  return result.split("").reverse().join("");
}

export { addBinary };

File: C++/src/add_binary.cpp
#include <add_binary.hpp>
using namespace std;

void trimLeadingZeros(string &s) {
  size_t firstOne = s.find('1');
  s = (firstOne == string::npos) ? "0" : s.substr(firstOne);
}

string Solution::addBinary(string a, string b) {
  trimLeadingZeros(a);
  trimLeadingZeros(b);

  int len_a = a.size();
  int len_b = b.size();

  if (len_a < len_b) {
    return addBinary(b, a);
  }

  int j = len_b - 1;
  int carry = 0;

  for (int i = len_a - 1; i >= 0; i--) {
    int bit1 = a[i] - '0';
    int sum = bit1 + carry;

    if (j >= 0) {
      int bit2 = b[j] - '0';
      sum += bit2;
      j--;
    }

    int bit = sum % 2;
    carry = sum / 2;

    a[i] = bit + '0';
  }

  if (carry > 0) {
    a = '1' + a;
  }

  return a;
}
